### DISCUSSION TITLE
Is stream interface fully  supported in  libvip

### DISCUSSION THREAD

**Discussion by joesong168** [libvips/libvips#3443] [General]
## Questions, enhancements, tips, etc.

As dzsave has supported stream interface to speed up content delivery, is other vips operation supported stream interface so that pipe operation is possible in command line, as 'vips xxx xxx | vips yyy yyy | vips zzz zzz'.

**Comment by jcupitt** [comment]
Hi again, 

You can join any operations like this, but you may see very high memory use, depending on what operations you connect. For example, a 90 degree rotate will have to buffer the entire image in memory.

**Reply by joesong168** [reply]
Great to hear that. What if the image has been shranked in previous vips operations so that even the entire image is in memory, it won't be too much. Cascade vips operations actually is very useful in image manipulation.

**Reply by jcupitt** [reply]
Connecting operations with pipes is *much*, *much* slower than using something like python and connecting them with images.

```python
image = pyvips.Image.new_from_file("k2.jpg", access="sequential")
image = image.fliplr().gaussblur(10).crop(10, 10, 100, 100)
image.write_to_file("x.v")
```

Streaming is useful for connecting the ends of a libvips pipeline to something other than a file or memory, but I wouldn't use it to join operations together.

**Comment by q3cpma** [comment]
Hello, I'll piggyback this issue to say I'm pretty confused concerning this streaming thing, which needs to be better documented (not even mentioned in https://www.libvips.org/API/current/using-cli.html), in my opinion.

The example in the changelog doesn't explain why I am not able to simply do
```
$ vips rot '[descriptor=0]' .png d90 <in.png >out.png
```
does this only work for operations taking a `VipsSource`? How should I change that rot example to make it work?

Thanks for reading.

**Reply by jcupitt** [reply]
You need:

```
$ vips rot stdin .png d90 <k2.png >out.png
```

The `[descriptor=0]` syntax is for operations which take a stream as input.

**Reply by q3cpma** [reply]
Huh, I see. Where can I read a more exhaustive description of the CLI? Where that stdin is mentioned, for example (is it a shorthand for /dev/stdin or does it work on Windows too?).

Thanks for the fast reply.

**Reply by jcupitt** [reply]
Yes, it'll work everywhere. `stdin` is a magic filename meaning the standard input.

The CLI is extremely simple and it's not really meant for production use, just for small odd jobs and for testing. I would use C# / Python / Ruby / php / C++ / etc. for anything important.

**Reply by q3cpma** [reply]
I guess I'll have to pony up for the full Python experience, indeed (or I'll make some Common Lisp bindings, eh). I'm sure it's not the first time you've heard it, but I was evaluating vips as a replacement for ImageMagick's quite complete CLI, hence my struggles =)

EDIT: ignore my mistake, .v doesn't work for some reason.

**Reply by q3cpma** [reply]
Or maybe .v/.vips not working can be considered a feature request or bug?

**Reply by jcupitt** [reply]
Ah sorry, I always forget, there's no streaming for the `.v` format for various technical reasons.

The very basic CLI was deliberate. ImageMagick has a full and pretty complex command evaluator, and I thought that scripting functionality was much better left to something like python. 

So instead, libvips aims to be very simple to bind -- it has (mostly) complete runtime introspection, so (assuming you have something like ffi available) you just need to fetch the argument list for an operation and marshal the objects to and from libvips representation. The core of pyvips (for example) is just a few 100 lines of code, with the rest being operator overloads, utilities, numpy integration, and so on. And because the binding is (in effect) generated at runtime, it's self-updating. When a new operator is added to libvips, it'll appear in pyvips automatically.

libvips is functional (no side effects, lazy) so it ought to be a good fit for CL.

**Comment by q3cpma** [comment]
Might be the wrong place, but here's a simplistic CLI I made myself, might serve some others:
```
#!/usr/bin/env python3

from sys import argv
import pyvips

oplist = [
    'pyvips.Image.new_from_source(pyvips.Source.new_from_descriptor(0), "")'
    if argv[1] == '-' else
    'pyvips.Image.new_from_file(argv[1])',

    *[s if s.endswith(')') else s + '()' for s in argv[2:-1]],

    'write_to_target(pyvips.vtarget.Target.new_to_descriptor(1), argv[-1])'
    if argv[-1].startswith('.') and argv[-1].count('.') == 1 else
    'write_to_file(argv[-1])'
]
eval('.'.join(oplist))
```
allows easy chaining with stdin and stdout support:
```
$ vips.py in.png out.png
$ vips.py in.png fliphor 'resize(.5)' out.png
$ vips.py - .png <in.png >out.png
```

**Reply by jcupitt** [reply]
Oh nice haha. Though `eval` always feels like cheating to me.

You can pass extra load options in filenames, so this works too:

```
vips.py in.png[access=sequential] fliphor 'resize(.5)' out.png
```

You could add something to pull any extra options from the `-` as well I guess, so you could do:

```
vips.py -[access=sequential] fliphor 'resize(.5)' .png[compression=9] < x.png > y.png
```

**Reply by q3cpma** [reply]
>Oh nice haha. Though eval always feels like cheating to me.

You do what you got to do when you have no macros =).

>You can pass extra load options in filenames, so this works too:

On that note, what do when you really have a file named `in.png[access=sequential]`? That's one of the reasons I'm not fond of automagic stuff. I guess you could have a keywork parameter `literal_path` or something to avoid interpretation.

**Reply by jcupitt** [reply]
The options are parsed by `new_from_file`, so you can call the loaders directly if you want a literal string loaded, eg. `x = pngload("in.png[access=sequential]")`.

Or write your own `new_from_file`, of course, it's pretty simple:

https://github.com/libvips/libvips/blob/master/libvips/iofuncs/image.c#L1916-L1942

The pyvips one:

https://github.com/libvips/pyvips/blob/master/pyvips/vimage.py#L340-L353

**Reply by q3cpma** [reply]
https://git.sr.ht/~q3cpma/scripts/tree/master/item/vips.py: last version with aforementioned `-[access=sequential]` support. Still thinking about CL bindings, since the pyvips module loading time is a bit high (0.1s here) for my taste.
